File: src/hrm.js
'use strict';

// Parser for the standard Bluetooth Heart Rate Measurement characteristic
// (0x2A37). The Polar H10 reports beat-to-beat RR intervals here, which is the
// robust path on Windows/WinRT and the high-quality source for RMSSD (R-wave
// detection is done by the H10 firmware itself).
//
// Accepts whatever the BLE layer hands us so the same parser is shared by both
// front-ends: a Node Buffer (noble, desktop) or a DataView (Capacitor BLE on
// Android, whose startNotifications callback delivers a DataView). A plain
// Uint8Array works too. All are normalised to a DataView and read little-endian.
//
// Flags byte (bit fields):
//   bit0 : HR value format (0 = uint8, 1 = uint16)
//   bit3 : Energy Expended present (uint16, skipped)
//   bit4 : RR-Interval(s) present (each uint16, units of 1/1024 s)
function parseHrm(input) {
  let dv;
  if (input instanceof DataView) {
    dv = input;
  } else if (input && input.buffer instanceof ArrayBuffer) {
    // Buffer / Uint8Array: respect byteOffset/length (Node Buffers are slices
    // of a shared, larger ArrayBuffer pool).
    dv = new DataView(input.buffer, input.byteOffset || 0, input.byteLength);
  } else {
    return { hr: null, rr: [] };
  }

  const len = dv.byteLength;
  if (len < 2) return { hr: null, rr: [] };

  const flags = dv.getUint8(0);
  let i = 1;

  let hr;
  if (flags & 0x01) {
    if (i + 2 > len) return { hr: null, rr: [] };
    hr = dv.getUint16(i, true);
    i += 2;
  } else {
    hr = dv.getUint8(i);
    i += 1;
  }

  if (flags & 0x08) i += 2; // energy expended -> skip

  const rr = [];
  if (flags & 0x10) {
    for (; i + 2 <= len; i += 2) {
      rr.push((dv.getUint16(i, true) / 1024) * 1000); // 1/1024 s -> ms
    }
  }

  return { hr, rr };
}

module.exports = { parseHrm };

File: app/android/app/src/main/java/dev/otake/rmssdh10n/hrv/Hrm.java
package dev.otake.rmssdh10n.hrv;

/**
 * Parser for the standard Bluetooth Heart Rate Measurement characteristic
 * (0x2A37). Java port of src/hrm.js — kept byte-for-byte equivalent so the
 * native service derives the same HR/RR the WebView pipeline does.
 *
 * Flags byte (bit fields):
 *   bit0 : HR value format (0 = uint8, 1 = uint16)
 *   bit3 : Energy Expended present (uint16, skipped)
 *   bit4 : RR-Interval(s) present (each uint16, units of 1/1024 s)
 * All multi-byte values are little-endian; bytes are read unsigned.
 */
public final class Hrm {
    /** HR in bpm (-1 when absent) and any RR intervals in ms. */
    public static final class Result {
        public final int hr;       // -1 if not present
        public final double[] rr;  // ms

        Result(int hr, double[] rr) { this.hr = hr; this.rr = rr; }
    }

    private static final double[] NO_RR = new double[0];

    public static Result parse(byte[] d) {
        if (d == null || d.length < 2) return new Result(-1, NO_RR);

        int flags = d[0] & 0xff;
        int i = 1;

        int hr;
        if ((flags & 0x01) != 0) {
            if (i + 2 > d.length) return new Result(-1, NO_RR);
            hr = (d[i] & 0xff) | ((d[i + 1] & 0xff) << 8);
            i += 2;
        } else {
            hr = d[i] & 0xff;
            i += 1;
        }

        if ((flags & 0x08) != 0) i += 2; // energy expended -> skip

        if ((flags & 0x10) != 0) {
            int n = (d.length - i) / 2;
            if (n > 0) {
                double[] rr = new double[n];
                int k = 0;
                for (; i + 2 <= d.length; i += 2) {
                    int raw = (d[i] & 0xff) | ((d[i + 1] & 0xff) << 8);
                    rr[k++] = raw / 1024.0 * 1000.0; // 1/1024 s -> ms
                }
                return new Result(hr, rr);
            }
        }
        return new Result(hr, NO_RR);
    }

    private Hrm() {}
}
